src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotPrimaryDataQueryInterface.java

```diff
@@ -1,39 +1,77 @@
 
   default StepResult prepareQueryAndCreateSnapshot(
       FlightContext context,
       Snapshot snapshot,
       SnapshotRequestModel snapshotReq,
-      DatasetService datasetService)
+      DatasetService datasetService,
+      SnapshotBuilderService snapshotBuilderService,
+      SnapshotRequestDao snapshotRequestDao,
+      SnapshotDao snapshotDao,
+      AuthenticatedUserRequest userReq)
       throws InterruptedException {
-    SnapshotRequestQueryModel snapshotQuerySpec = snapshotReq.getContents().get(0).getQuerySpec();
+    AssetSpecification assetSpecification;
+    String sqlQuery;
+    Instant createdAt;
 
-    Query query = Query.parse(snapshotQuerySpec.getQuery());
-    String datasetName = query.getDatasetName();
-    Dataset dataset = datasetService.retrieveByName(datasetName);
-    AssetSpecification assetSpecification =
-        retrieveAssetSpecification(dataset, snapshotQuerySpec.getAssetName());
-    validateRootTable(query, assetSpecification);
-    String sqlQuery = translateQuery(query, dataset);
+    if (snapshotReq
+        .getContents()
+        .get(0)
+        .getMode()
+        .equals(SnapshotRequestContentsModel.ModeEnum.BYREQUESTID)) {
+      UUID accessRequestId =
+          snapshotReq.getContents().get(0).getRequestIdSpec().getSnapshotRequestId();
+      SnapshotAccessRequestResponse accessRequest = snapshotRequestDao.getById(accessRequestId);
+
+      UUID dataReleaseSnapshotId = accessRequest.getSourceSnapshotId();
+      Snapshot dataReleaseSnapshot = snapshotDao.retrieveSnapshot(dataReleaseSnapshotId);
+      // get the underlying dataset for the snapshot
+      if (dataReleaseSnapshot.getSnapshotSources().isEmpty()) {
+        throw new IllegalArgumentException("Snapshot does not have a source dataset");
+      }
+      Dataset dataset = dataReleaseSnapshot.getSnapshotSources().get(0).getDataset();
+      // gets pre-existing asset on the dataset
+      // TODO: create custom asset
+      assetSpecification = dataset.getAssetSpecificationByName("concept_asset").orElseThrow();
+      sqlQuery =
+          snapshotBuilderService.generateRowIdQuery(accessRequest, dataReleaseSnapshot, userReq);
+      createdAt = dataReleaseSnapshot.getCreatedDate();
+    } else {
+      SnapshotRequestQueryModel snapshotQuerySpec = snapshotReq.getContents().get(0).getQuerySpec();
+
+      Query query = Query.parse(snapshotQuerySpec.getQuery());
+      String datasetName = query.getDatasetName();
+      Dataset dataset = datasetService.retrieveByName(datasetName);
+      assetSpecification = retrieveAssetSpecification(dataset, snapshotQuerySpec.getAssetName());
+      validateRootTable(query, assetSpecification);
+      sqlQuery = translateQuery(query, dataset);
+      createdAt = CommonFlightUtils.getCreatedAt(context);
+    }
 
-    Instant createdAt = CommonFlightUtils.getCreatedAt(context);
     return createSnapshotPrimaryData(context, assetSpecification, snapshot, sqlQuery, createdAt);
```

Comment:
I would rather duplicate the return statement then instantiate null variables for `assetSpecification`, `sqlQuery`, and `createdAt`. I think that would be clearer, and also make it less error prone (I could see us adding another case here, but forgetting to instantiate those variables.

If we use a switch statement, we could also move to a switch expression and that way just return the switch.